Anesthetic tanks buff surgery speed

thatoneplebeian opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->
Anesthetic tanks add 20% to surgery speed when used as the target's internals. This stacks with the bonus from reagents.

## Why It's Good For The Game

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->
Anesthetic tanks, as they are currently, are essentially useless items. They sit in the closet all shift and no doctor ever pulls one out. Adding a benefit to using them would not only give them purpose but also provide a cover for antagonists who want to use the tanks for their own devious plans.

## Changelog
:cl: ThatOnePlebeian
balance: Being anesthetized (from N2O gas) boosts surgery speed.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Commit:
Anesthetic tanks buff surgery speed

Anesthetic tanks add 20% to surgery speed when used as the target's internals

Commit:
Merge remote-tracking branch 'upstream/master' into anestheticbuff

Jack7D1 commented:
Reward behavior you want. Don't punish behavior you don't want.
This PR is the very embodiment of good practice

kevinz000 commented on code/modules/surgery/surgery_step.dm:
```diff
@@ -82,7 +82,7 @@
 		implement_speed_mod = implements[implement_type] / 100.0
 
-	speed_mod /= (get_location_modifier(target) * (1 + surgery.speed_modifier) * implement_speed_mod)
+	speed_mod /= (get_location_modifier(target) * (1 + get_internals_modifier(target)) * (1 + surgery.speed_modifier) * implement_speed_mod)
```
Is the goal here to reward only those using proper anesthetic tanks from medbay or being sedated during surgery?

If it's the latter and the goal is to reward being vulnerable during surgery, it would be much better to check for sources of being knocked out than to do a snowflake check for a singular group of tank typepaths.

kevinz000 commented on code/modules/surgery/surgery_helpers.dm:
```diff
@@ -130,6 +130,16 @@
 	else
 		return 0.5
 
+/proc/get_internals_modifier(mob/M)
```
If you really want to go the path of forcing *anesthetic tank only* for this, this proc really should be more descriptive on what it's doing, as get_internals_modifier doesn't say what said modifier is *for*.

thatoneplebeian replied:
Done, thank you!

kevinz000 commented on code/modules/surgery/surgery_helpers.dm:
```diff
@@ -130,6 +130,16 @@
+	var/mob/living/carbon/human/H = M
+	if (!istype(H))
+		return 0.0
+	else if (istype(H.internal,/obj/item/tank/internals/anesthetic))
```
And this allows for them to just fill it up with oxygen and release at 16 kpa, it'll count as being internal. They just won't be knocked out.

thatoneplebeian replied:
Ah, absolutely. I will get working on implementing a check when I have some free time. People refilling anesthetic tanks is an unlikely scenario but it is important to cover all possibilities. I also just realized that this would work if the tank was empty. As I become more familiar with the byond environment and the tgstation code I should learn to avoid simple mistakes like this. Thank you for your feedback!

ExcessiveUseOfCobblestone replied:
replace with a sleeping/unconscious check? 

The tank should be giving you enough to go unconscious.

Commit:
Renamed proc to be more specific

Renamed proc from "get_internals_modifier" to "get_anesthetic_modifier" to better show that it is only relevant to anesthetic tanks.

Tlaltecuhtli commented:
this is dumb:
- the 20% speed is lost by the 30s you spend putting the n2o tank on person for most surgeries
- corpses get this buff
- should be something more generic like if person is unconscious so you can use sleepy chems for anesthesia

XDTM commented:
Yeah, the check should be based on unconsciousness. If that is too broad, then an anesthetic trait given by N2O or morphine on top of the unconsciousness.

Jared-Fogle commented:
Just checking for unconsciousness is a pretty big buff, most of the big work I do is done while they're still unconscious after being revived.

ExcessiveUseOfCobblestone commented:
> Reward behavior you want. Don't punish behavior you don't want.
> This PR is the very embodiment of good practice

surgeries are now 20% slower if the user doesn't use an anesthetic tank is an equally accurate statement.

The only difference between this in the code is that now 20% has to be mentally shaved off in certain situations when you're trying to come up with an appropriate time instead of the time you set for the surgery being the "the stars align" number. The perfect option really would be to code it like a punishment but market it as a reward but that is a discussion best left out of this pr hehe.

XDTM commented:
> 
> 
> > Reward behavior you want. Don't punish behavior you don't want.
> > This PR is the very embodiment of good practice
> 
> surgeries are now 20% slower if the user doesn't use an anesthetic tank is an equally accurate statement.
> 
> The only difference between this in the code is that now 20% has to be mentally shaved off in certain situations when you're trying to come up with an appropriate time instead of the time you set for the surgery being the "the stars align" number. The perfect option really would be to code it like a punishment (for coders) but market it as a reward (for players) but that is a discussion best left out of this pr hehe.

I'm pretty sure this is a widely used technique in game design, a lot of penalties can simply be worded as bonuses in the opposite situation and have a much better player reception.

ExcessiveUseOfCobblestone commented:
> 
> I'm pretty sure this is a widely used technique in game design, a lot of penalties can simply be worded as bonuses in the opposite situation and have a much better player reception.

correct! Wurm has a good example of this (probably taken from WoW) where if you "sleep" aka log off in a bed you can gain a boost in xp gain when you return. This could also be phrased as if you constantly play to the point where you don't log off, you are losing the ability to get this xp bonus.

thatoneplebeian commented:
> 
> 
> Yeah, the check should be based on unconsciousness. If that is too broad, then an anesthetic trait given by N2O or morphine on top of the unconsciousness.

>
>
> replace with a sleeping/unconscious check?

Thank you all for your suggestions! I'm not sure about implementing a broad unconsciousness check, as Jared-Fogle mentioned, patients are unconscious nearly half the time anyways, and the general check would undermine what I had actually hoped to accomplish by implementing this (getting people to use the tanks). 
XDTM's idea of adding a trait given by N2O when inhaled is definitely a much better way of implementing the change; I will not have to write some "snowflake" code as the trait could be given to more gases/reagents in the future.
The anesthetic tanks would then provide a ready-to-use way to get the bonus (or avoid the penalty, whichever way you look at it), while still allowing for the bonus to be obtained when the tanks are not available.

evilcyborgmrt commented:
This is great but people will still spaz out the moment a doctor grabs an anaesthetic tank

anconfuzedrock commented:
just a reminder you can perform a sleep action in the ic tab

Pugie commented:
Patient being KOd should increase surgery speed more and it should be adjusted that being awake means steps are constantly failed, this gives more importance to high medical skill as legendary medics will be capable of performing emergency surgeries on people who are awake.

ExcessiveUseOfCobblestone commented:
you could just have it check if not conscious and not in crit or something, a new trait just to make it exclusive to literally 1 interaction makes very little sense to me even though I understand your only goal here is for the tank to get more use.

thatoneplebeian commented:
What @Pugie commented is the only way I can see a broad implementation of this actually working. It would have to make performing surgeries on awake people difficult to the point where it is worth putting them under. 
This seems to me like a more "realistic" implementation but not one that would make playing medical more enjoyable.

I currently have an implementation working that uses status effects. Like @ExcessiveUseOfCobblestone said, it seems odd to have a new effect for just one interaction, but the idea here is that it is possible for future changes to make use of the effect.
I should make the commit pretty soon so the changes can be reviewed.

>
>
> This is great but people will still spaz out the moment a doctor grabs and anesthetic tank

Wrangling a mask and tank onto your half-dead patient as they try helplessly to crawl away screaming "no gas!" is just part of the fun, right?

Commit:
Anesthetic rework

Anesthetized is now a status effect, currently only obtainable by breathing in N2O.
Anesthetized patients will boost surgery speed by 20%.

optimumtact commented:
why is it so high by default?

tmtmtl30 commented:
not sure i like making surgery near-impossible when patients are conscious, since being required to totally give up the ability to play the game in order to get healed in any reasonable amount of time isn't very fun. let's make surgery more important to healing -> let's make surgery incredibly unattractive to any patient feels like a _weird_ route to take.

Cyberboss commented on code/modules/surgery/organs/lungs.dm:
```diff
@@ -243,6 +243,7 @@
 			H.Unconscious(60) // 60 gives them one second to wake up and run away a bit!
 			if(SA_pp > SA_sleep_min) // Enough to make us sleep as well
 				H.Sleeping(max(H.AmountSleeping() + 40, 200))
+				H.Anesthetized(max(H.AmountAnesthetized() + 40, 200)) //N2O is an anesthetic agent
```
I feel like Anesthetized should conditionally apply the Sleeping effect

silicons replied:
isn't this way better as a trait or something? why does anesthetics even need a status effect?

silicons replied:
status effects are way heavier than traits and unless you want something where it slowly wears off after you stop using it it's a waste (and yes i can see you doing that here but uh if they're conscious they should probably feel pain tbh)

optimumtact commented:
@tmtmtl30 this doesn't change default surgery? It buffs unconcious patients speed.

Cyberboss commented:
Uhh, you sure about that oranges? It looks ok to me.

On Fri, Jun 19, 2020, 9:01 PM oranges <notifications@github.com> wrote:

> @tmtmtl30 <https://github.com/tmtmtl30> this doesn't change default
> surgery? It buffs unconcious patients speed.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/tgstation/tgstation/pull/51715#issuecomment-646911204>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB6LA6VEYAOXNU3ZSSTVAWDRXQC45ANCNFSM4OBE4URQ>
> .
>

optimumtact commented:
that's not what I meant

Cyberboss commented:
Somehow i did not get an email for any comment except yours

ExcessiveUseOfCobblestone commented:
I'm not terribly sold on this only because surgery is already a bore (especially on the patient).

Is there any allure to this besides the realism argument that people want to afk-freeze individiuals?

Cyberboss commented:
@ExcessiveUseOfCobblestone Basically no one agrees to be anesthetized for surgery currently. It really needs to help the receiving player somehow so it can become a more used mechanic as well as an option for traitor docs.

ExcessiveUseOfCobblestone commented:
Right, it needs to be desirable to the player.

I guess my concern is that even if you make it desirable to the player from a cost:benefit pov, the reality is the mechanic encourages the player to do something short of alt-tabbing.

From a medical pov, I think there is a delicate balance because you have to appease both doctors (what I tend to focus on) AND patients (sleeper/trekchem era). As someone who has moved towards surgery, I will be the first to tell you it is not terribly enjoyable for either side even if it does a lot of things right for balance.

Forcing your **living** patients to be in a situation where tabbing out of the game watching runtime videos on y*utube is more tasteful than being tabbed into the game because of a number slider, regardless of how great that number slider is, feels like a design failing to me.

Shadowflame909 commented:
What cobby said

Dreaming needs a buff.

Dreaming minigame that doesn't greatly raise server CPU when.

tmtmtl30 commented:
@optimumtact was responding to thatoneplebeian, not this PR, this one is fine

tmtmtl30 commented:
(also, cryo already exists and offers a similar trade (definite healing for a period of time where you're totally vulnerable and out of the loop), and i think it's fine there. it can be pretty tense)

ExcessiveUseOfCobblestone commented:
"It exists here too" is a bit of a distraction from the discussion of whether or not that tradeoff is good design.

kriskog commented:
Merge conflict.

Ideally, I think this needs a bit more fleshing out. Instead of looking at what internals someone is wearing, being unable to control your character should make the surgery faster on you. This would also apply to dead people, for reasons. But to prevent cheesing it with the sleep verb only forced unconsciousness should do this, and this is where the fleshing out comes in. Perhaps forcing waking up when damage is applied if sleep is not forced. It would open up the possibility of drugging patients by other means than internals as well.

WarlockD commented:
Could it also midigate o2 damnage that comes right after revival and maybe act as a CPR?   If you get that no breathing virus is hard to get someone out of a simple crit

kriskog commented:
@ a maintainer once you are ready to work on this again
